src/core/qgsarrayutils.cpp

```diff
@@ -0,0 +1,94 @@
+/***************************************************************************
+                                  qgsarrayutils.h
+                              ---------------------
+    begin                : July 2019
+    copyright            : (C) 2019 by David Signer
+    email                : david at opengis dot ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsarrayutils.h"
+#include <QDebug>
+#include <nlohmann/json.hpp>
+using json = nlohmann::json;
+
+QVariantList QgsArrayUtils::parse( const QString &string )
```

Comment:
Should this be called `PgArrayUtils` since it's a very particular array style? (And the only other array style in use is already handled by the `JsonUtils`, so making it highly configurable is of limited use at the moment for what I can see).

Reply:
It's this general name in case we want to extend the utils to other formats. But I can rename it to `PgArrayUtils` if you want... I'm not fixed on that.

Reply:
Right now it's called `QgsArrayUtils::parse`, so it sounds like it's generic but it's implemented specific. I think either the method name or the class name should reflect the specifity.

Reply:
I agree. I vote for the method name...

Reply:
After thinking about. In case we replace all the postgres array parse functions with this it's maybe better to call it qgsPgArrayUtils...

Reply:
> After thinking about. In case we replace all the postgres array parse functions with this it's maybe better to call it qgsPgArrayUtils...

Is there any reason for not using the actual postgres provider code? Do you plan to support the multidimensional array or provider specific type (like a boolean array)?

Reply:
In case there is a function parsing and building postgres arrays it would make sense when the postgres provider uses them too, right? But when the postgres provider uses functions there, it would probably make sense to move all parsing/building functions there too. But I'm not sure if I want to make this massive change. I'm checking out what would have to be done then...

Reply:
I'm quite sure that both postgresprovider and widget that need to parse array needs to call the same code to parsearray. 

Possibly, the one you are currently proposing. But there is already a bunch of methods in postgres provider (getNextString, parseArray, parseStringArray...) that already parse array and deals with some issues you are facing to (quotes, escaped character, multidimensionnal array...). 

Why not copying them in a QgsArrayUtils class instead of rewriting a new parser? But maybe I miss something.

Reply:
Copying them is anyway a possibility I am heading to. But maybe in the future we can reduce it to one function used everywhere.